src/main/java/org/checkerframework/checker/androidresource/AndroidResourceVisitor.java

```diff
@@ -0,0 +1,104 @@
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.framework.source.Result;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+import javax.lang.model.element.AnnotationMirror;
+
+public class AndroidResourceVisitor extends BaseTypeVisitor<AndroidResourceAnnotatedTypeFactory> {
+
+    /**
+     * The @ResourceTop annotation.
+     */
+    protected final AnnotationMirror RESOURCE_TOP =
+            AnnotationBuilder.fromClass(elements, ResourceTop.class);
+
+    /**
+     * The @ResourceBottom annotation.
+     */
+    protected final AnnotationMirror RESOURCE_BOTTOM =
+            AnnotationBuilder.fromClass(elements, ResourceBottom.class);
+
+    public AndroidResourceVisitor(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    /**
+     * Gives a warning whenever operands in a binary assignment have a annotation of type @XXXRes.
+     */
+    @Override
+    public Void visitBinary(BinaryTree node, Void aVoid) {
+
+        ExpressionTree leftOperand = node.getLeftOperand();
+        ExpressionTree rightOperand = node.getRightOperand();
+        AnnotatedTypeMirror leftOperandType = atypeFactory.getAnnotatedType(leftOperand);
+        AnnotatedTypeMirror rightOperandType = atypeFactory.getAnnotatedType(rightOperand);
+
+        Kind kind = node.getKind();
+
+//        System.out.println(node.toString() + ", " + node.getKind() + "\n");
+
+        switch (kind) {
+            case EQUAL_TO:
+                break;
+            case NOT_EQUAL_TO:
+                break;
+            default:
+                if (checkAnnotatedTypeHasResOrContainerAnnotations(leftOperandType) || checkAnnotatedTypeHasResOrContainerAnnotations(rightOperandType)) {
+                    checker.report(Result.warning("binary.operation.not.allowed", kind), node);
+                }
+        }
+
+        return super.visitBinary(node, aVoid);
+    }
+
+    /**
+     * Gives a warning whenever expression/variable in a compound assignment have a annotation of type @XXXRes.
+     */
+    @Override
+    public Void visitCompoundAssignment(CompoundAssignmentTree node, Void p) {
+
+        ExpressionTree expression = node.getExpression();
+        ExpressionTree variable = node.getVariable();
+        AnnotatedTypeMirror expressionType = atypeFactory.getAnnotatedType(expression);
+        AnnotatedTypeMirror variableType = atypeFactory.getAnnotatedType(variable);
+
+        Kind kind = node.getKind();
+
+//        System.out.println(node.toString() + ", " + node.getKind() + "\n");
+
+        if (checkAnnotatedTypeHasResOrContainerAnnotations(expressionType) || checkAnnotatedTypeHasResOrContainerAnnotations(variableType)) {
+            checker.report(Result.warning("compound.assignment.not.allowed", kind), node);

```

Comment:
The error message would be more useful if it contained the specific part that is incorrect and an explanation of what is wrong. The error message just says  `Resource annotation` instead of what the concrete problem is.

Reply:
I have added a more verbose error message. Please check.